src/core/rowindex_array.cc

```diff
@@ -314,40 +313,47 @@ void ArrayRowIndexImpl::compactify()
   _resize_data();
 }
 
+template<typename TI>
+void get_next_index_to_skip(const TI* inputs, size_t& j)
+{
+  while (inputs[j] == inputs[j+1]) {
+    ++j;
+  }
+  ++j;
+}
 
 template <typename TI, typename TO>
 RowIndexImpl* ArrayRowIndexImpl::negate_impl(size_t nrows) const
 {
   auto inputs = static_cast<const TI*>(buf_.rptr());
-  size_t newsize = nrows - length;
+  size_t newsize = nrows;
   size_t inpsize = length;
   Buffer outbuf = Buffer::mem(newsize * sizeof(TO));
   auto outputs = static_cast<TO*>(outbuf.xptr());
   TO orows = static_cast<TO>(nrows);
 
   TO next_index_to_skip = static_cast<TO>(inputs[0]);
-  size_t j = 1;  // next index to read from the `inputs` array
+  size_t count_indices_to_skip = 0;
+  size_t j = 0;  // index to read from the `inputs` array
   size_t k = 0;  // next index to write into the `outputs` array
   for (TO i = 0; i < orows; ++i) {
     if (i == next_index_to_skip) {
-      next_index_to_skip =
-        j < inpsize? static_cast<TO>(inputs[j++]) : orows;
-      if (next_index_to_skip <= i) {
-        throw ValueError() << "Cannot invert RowIndex which is not sorted";
-      }
+      ++count_indices_to_skip;
+      get_next_index_to_skip(inputs, j);
```

Comment:
The name of this function does not reflect what it does. `get_smth()` ought to be returning the value `smth`. However, in your case the function is about advancing `j` to the correct position. So i'd suggest to either (1) rename the function, or (2) inline it, or (3) move calculation of `next_index_to_skip` into the function and return that value.

Reply:
I agree, no need to move it into a separate function. I have brought the computation back into the calling function.